crates/storage/provider/src/providers/database/provider.rs

```
@@ -1043,21 +1043,31 @@ impl<TX: DbTx> BlockReader for DatabaseProvider<TX> {
     /// will return None.
     fn block_with_senders(
         &self,
-        block_number: BlockNumber,
+        block_number: BlockHashOrNumber,
         transaction_kind: TransactionVariant,
     ) -> RethResult<Option<BlockWithSenders>> {
-        let Some(header) = self.header_by_number(block_number)? else { return Ok(None) };
+        let block_num: BlockNumber = match block_number {
+            BlockHashOrNumber::Hash(hash) => {
+                let res = match self.block_number(hash)? {
+                    Some(num) => Ok(num),
+                    None => Err("Block number not found for given hash"),
```

this should return the appropriate error